static/app/components/replays/utils.tsx

```diff
@@ -28,3 +30,90 @@ export function formatTime(ms: number): string {
+ * represent an hour.
+ *
+ * @param duration The amount of time that we need to chop up into even sections
+ * @param width Total width available, pixels
+ * @param minWidth Minimum space for each column, pixels. Ex: So we can show formatted time like `1:00:00` between major ticks
+ * @returns
+ */
+export function countColumns(duration: number, width: number, minWidth: number = 50) {
+  let maxCols = Math.floor(width / minWidth);
+  const remainder = duration - maxCols * width > 0 ? 1 : 0;
+  maxCols -= remainder;
+
+  // List of all the possible time granularities to display
+  // We could generate the list, which is basically a version of fizzbuzz, hard-coding is quicker.
+  const timeOptions = [
+    1 * HOUR,
+    30 * MINUTE,
+    20 * MINUTE,
+    15 * MINUTE,
+    10 * MINUTE,
+    5 * MINUTE,
+    2 * MINUTE,
+    1 * MINUTE,
+    30 * SECOND,
+    10 * SECOND,
+    5 * SECOND,
+    1 * SECOND,
+  ];
+
+  const timeBasedCols = timeOptions.reduce<Map<number, number>>((map, time) => {
+    map.set(time, Math.floor(duration / time));
+    return map;
+  }, new Map());
+
+  const [timespan, cols] = Array.from(timeBasedCols.entries())
+    .filter(([_span, c]) => c <= maxCols) // Filter for any valid timespan option where all ticks would fit
+    .reduce((best, next) => (next[1] > best[1] ? next : best), [0, 0]); // select the timespan option with the most ticks
+
+  const remaining = (duration - timespan * cols) / timespan;
+  return {timespan, cols, remaining};
+}
+
+/**
+ * Group Crumbs for display along the timeline.
+ *
+ * The timeline is broken down into columns (aka buckets, or time-slices).
+ * Columns translate to a fixed width on the screen, to prevent side-scrolling.
+ *
+ * This function groups crumbs into columns based on the number of columns available
+ * and the timestamp of the crumb.
+ */
+export function getCrumbsByColumn(crumbs: Crumb[], totalColumns: number) {
+  const startTime = crumbs[0]?.timestamp;
+  const endTime = crumbs[crumbs.length - 1]?.timestamp;
+
+  // If there is only one crumb then we cannot do the math, return it in the first column
+  if (crumbs.length === 1 || startTime === endTime) {
+    return new Map([[0, crumbs]]);
+  }
+
+  const startMilliSeconds = +new Date(String(startTime));
+  const endMilliSeconds = +new Date(String(endTime));
+
+  const duration = endMilliSeconds - startMilliSeconds;
+  const safeDuration = isNaN(duration) ? 1 : duration;
+
+  const eventsByCol = crumbs.reduce<Map<number, Crumb[]>>((map, breadcrumb) => {
+    const {timestamp} = breadcrumb;
+    const timestampMilliSeconds = +new Date(String(timestamp));
+    const sinceStart = isNaN(timestampMilliSeconds)
+      ? 0
+      : timestampMilliSeconds - startMilliSeconds;
+    const column = Math.floor((sinceStart / safeDuration) * (totalColumns - 1)) + 1;
+
+    map.set(column, [...Array.from(map.get(column) || []), breadcrumb]);
+    return map;
+  }, new Map());
```

Comment:
With my first read-through, this kind of caught me off-guard using `reduce()` like this to initialize a Map, and same feelings the second time too. No strong preference her, but would it be a tad easier to read if we mapped over crumbs and returned a list of tuples to pass to Map's constructor?

Reply:
that sounds really good. It started with an object which was fine but wanted the key to be a `number` so it all evolved. The place that consumes everything uses entries as well, so I could just skip the Map altogether.

Reply:
So, I couldn't sort out how to get rid of the map inside here actually. The `Map` is used to accumulate the crumbs by column. If the crumbs are guaranteed to be ordered we could do it a little differently... but instead I think i made it clearer with an initial `.map()` call followed by the `.reduce()` call that does the grouping.